src/coord/src/coord.rs

```diff
@@ -393,41 +406,65 @@ impl<C> Coordinator<C>
 where
     C: dataflow_types::client::Client + 'static,
 {
-    /// Assign a timestamp for a read.
-    fn get_read_ts(&mut self) -> Timestamp {
-        let ts = self.get_ts();
-        self.last_op_was_read = true;
-        self.read_lower_bound = ts;
-        ts
+    /// Assign a timestamp for a read from a local input. Reads following writes
+    /// must be at a time >= the write's timestamp; we choose "equal to" for
+    /// simplicity's sake and to open as few new timestamps as possible.
+    fn get_local_read_ts(&mut self) -> Timestamp {
+        // If you have pending writes...
+        if self.writes_at_open_ts {
+            // You will need to read those writes, which happened at the last
+            // known open time. This also means you will need to advance to
+            // those writes, i.e. close over `last_open_local_ts`.
+            self.read_at_offset_from_open_local = 0;
+            self.writes_at_open_ts = false;
+        }
+
+        self.last_open_local_ts - self.read_at_offset_from_open_local
     }
 
-    /// Assign a timestamp for a write. Writes following reads must ensure that they are assigned a
-    /// strictly larger timestamp to ensure they are not visible to any real-time earlier reads.
-    fn get_write_ts(&mut self) -> Timestamp {
-        let ts = if self.last_op_was_read {
-            self.last_op_was_read = false;
-            cmp::max(self.get_ts(), self.read_lower_bound + 1)
-        } else {
-            self.get_ts()
-        };
-        self.read_lower_bound = cmp::max(ts, self.closed_up_to);
-        self.read_lower_bound
+    /// Assign a timestamp for a write to a local input. Writes following reads
+    /// must ensure that they are assigned a strictly larger timestamp to ensure
+    /// they are not visible to any real-time earlier reads.
+    fn get_local_write_ts(&mut self) -> Timestamp {
+        // This assert is valid because:
+        // - Whenever a write precedes a read, the read sets
+        //   `read_at_offset_from_open_local = 0`, which will advance the
+        //   `last_open_local_ts`.
+        // - Reads always enter a greenthread, letting the Coordinator see that
```

Comment:
I don't understand the claim that reads always enter a greenthread? Are you claiming that because table reads (previously) always had to be put into the pending peeks lists in dataflow? I think with this new code that's no longer true.

I think the threadedness of the read doesn't matter so much, and it's more that the coordinator always checks at the end of each message if it should advance local inputs, so that coord will have always done the right thing immediately after any read.

For example, even if the peek was immediately ready, coord would never process that because peek ready is another message it'd have to dequeue, but it'll do the advance check first.

Reply:
Ah, I mangled the explanation with mentioning a greenthread because it's untrue in the case of `implement_fast_path_peek`.

The point I am ultimately making is that whenever the coordinator performs a read, it has a chance to check `self.read_at_offset_from_open_local == 0` (i.e. needs to open a new timestamp) before executing any writes. In the case of executing `ReadThenWrite` plans, this is facilitated by the read entering a greenthread, but in the case of regular peeks, it occurs naturally. I'll amend the comment here to make it more general.